Compiling on OS X

I have had a lot of trouble compiling on OS X using the method described in the readme.

I have got it working with the following instead (requires g++-4.7).

```
cd /tmp
git clone https://github.com/v8/v8.git
cd v8
make dependencies

export CXX=/usr/local/bin/g++-4.7
export LINK=/usr/local/bin/g++-4.7
make native component=shared_library snapshot=off -j8

# Important to turn snapshot off

sudo mkdir -p /tmp/v8-install/lib /tmp/v8-install/include

# Copy the dylib as an "so" doesn't get built
sudo cp out/native/lib*.dylib /tmp/v8-install/lib/
sudo cp include/v8* /tmp/v8-install/include

cd /tmp
git clone https://github.com/preillyme/v8js.git
cd v8js
phpize

# Need to configure with D_GLIBCXX_USE_NANOSLEEP
./configure --with-v8js=/tmp/v8-install CXXFLAGS="-D_GLIBCXX_USE_NANOSLEEP"
make
```

I also needed to remove the "-Wl,-rpath" flags in the config.m4 file:

```
dnl $Id$

PHP_ARG_WITH(v8js, for V8 Javascript Engine,
[  --with-v8js           Include V8 JavaScript Engine])

if test "$PHP_V8JS" != "no"; then
  SEARCH_PATH="/usr/local /usr"
  SEARCH_FOR="include/v8.h"
  
  if test -r $PHP_V8JS/$SEARCH_FOR; then
    LDFLAGS="$LDFLAGS"
    V8_DIR=$PHP_V8JS
  else
    AC_MSG_CHECKING([for V8 files in default path])
    for i in $SEARCH_PATH ; do
      if test -r $i/$SEARCH_FOR; then
        V8_DIR=$i
        AC_MSG_RESULT(found in $i)
      fi
    done
  fi

  if test -z "$V8_DIR"; then
    AC_MSG_RESULT([not found])
    AC_MSG_ERROR([Please reinstall the v8 distribution])
  fi

  PHP_ADD_INCLUDE($V8_DIR/include)
  PHP_ADD_LIBRARY_WITH_PATH(v8, $V8_DIR/$PHP_LIBDIR, V8JS_SHARED_LIBADD)
  PHP_SUBST(V8JS_SHARED_LIBADD)
  PHP_REQUIRE_CXX()

  old_LIBS=$LIBS
  old_LDFLAGS=$LDFLAGS
  old_CPPFLAGS=$CPPFLAGS
  LDFLAGS="-L$V8_DIR/$PHP_LIBDIR"
  LIBS=-lv8
  CPPFLAGS=-I$V8_DIR/include
  AC_LANG_SAVE
  AC_LANG_CPLUSPLUS

  AC_CACHE_CHECK(for V8 version, ac_cv_v8_version, [
AC_TRY_RUN([#include <v8.h>
#include <iostream>
#include <fstream>
using namespace std;

int main ()
{
	ofstream testfile ("conftestval");
	if (testfile.is_open()) {
		testfile << v8::V8::GetVersion();
		testfile << "\n";
		testfile.close();
		return 0;
	}
	return 1;
}], [ac_cv_v8_version=`cat ./conftestval|awk '{print $1}'`], [ac_cv_v8_version=NONE], [ac_cv_v8_version=NONE])
])

  if test "$ac_cv_v8_version" != "NONE"; then
    ac_IFS=$IFS
    IFS=.
    set $ac_cv_v8_version
    IFS=$ac_IFS
    V8_API_VERSION=`expr [$]1 \* 1000000 + [$]2 \* 1000 + [$]3`
    if test "$V8_API_VERSION" -lt 3021012 ; then
       AC_MSG_ERROR([libv8 must be version 3.21.12 or greater])
    fi
    AC_DEFINE_UNQUOTED([PHP_V8_API_VERSION], $V8_API_VERSION, [ ])
    AC_DEFINE_UNQUOTED([PHP_V8_VERSION], "$ac_cv_v8_version", [ ])
  else
    AC_MSG_ERROR([could not determine libv8 version])
  fi

  AC_CACHE_CHECK(for debuggersupport in v8, ac_cv_v8_debuggersupport, [
AC_LINK_IFELSE([AC_LANG_PROGRAM([[#include <v8-debug.h>]],
                               [[v8::Debug::DisableAgent()]])],
    [ac_cv_v8_debuggersupport=yes],
    [ac_cv_v8_debuggersupport=no])
])

  if test "$ac_cv_v8_debuggersupport" = "yes"; then
    AC_DEFINE([ENABLE_DEBUGGER_SUPPORT], [1], [Enable debugger support in V8Js])
  fi

  AC_LANG_RESTORE
  LIBS=$old_LIBS
  LDFLAGS=$old_LDFLAGS
  CPPFLAGS=$old_CPPFLAGS


  AC_CACHE_CHECK(for C standard version, ac_cv_v8_cstd, [
    ac_cv_v8_cstd="c++11"
    old_CPPFLAGS=$CPPFLAGS
    AC_LANG_PUSH([C++])
    CPPFLAGS="-std="$ac_cv_v8_cstd
    AC_TRY_RUN([int main() { return 0; }],[],[ac_cv_v8_cstd="c++0x"],[])
    AC_LANG_POP([C++])
    CPPFLAGS=$old_CPPFLAGS
  ]);
  
  PHP_NEW_EXTENSION(v8js, v8js.cc v8js_convert.cc v8js_methods.cc v8js_variables.cc v8js_commonjs.cc, $ext_shared, , "-std="$ac_cv_v8_cstd)

  PHP_ADD_MAKEFILE_FRAGMENT
fi
```

cant build on Maverics too.. (

```
➜  v8js git:(master) pwd
/Users/fro/build/v8js
➜  v8js git:(master) ./configure --with-v8js=/Users/fro/build/v8 CXXFLAGS="-D_GLIBCXX_USE_NANOSLEEP"
checking for grep that handles long lines and -e... /usr/local/bin/ggrep
checking for egrep... /usr/local/bin/ggrep -E
checking for a sed that does not truncate output... /usr/local/bin/gsed
checking for cc... cc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether cc accepts -g... yes
checking for cc option to accept ISO C89... none needed
checking how to run the C preprocessor... cc -E
checking for icc... no
checking for suncc... no
checking whether cc understands -c and -o together... yes
checking for system library directory... lib
checking if compiler supports -R... no
checking if compiler supports -Wl,-rpath,... yes
checking build system type... x86_64-apple-darwin14.1.0
checking host system type... x86_64-apple-darwin14.1.0
checking target system type... x86_64-apple-darwin14.1.0
checking for PHP prefix... /usr/local/Cellar/php56/5.6.5
checking for PHP includes... -I/usr/local/Cellar/php56/5.6.5/include/php -I/usr/local/Cellar/php56/5.6.5/include/php/main -I/usr/local/Cellar/php56/5.6.5/include/php/TSRM -I/usr/local/Cellar/php56/5.6.5/include/php/Zend -I/usr/local/Cellar/php56/5.6.5/include/php/ext -I/usr/local/Cellar/php56/5.6.5/include/php/ext/date/lib
checking for PHP extension directory... /usr/local/Cellar/php56/5.6.5/lib/php/extensions/debug-non-zts-20131226
checking for PHP installed headers prefix... /usr/local/Cellar/php56/5.6.5/include/php
checking if debug is enabled... yes
checking if zts is enabled... yes
checking for re2c... no
configure: WARNING: You will need re2c 0.13.4 or later if you want to regenerate PHP parsers.
checking for gawk... no
checking for nawk... no
checking for awk... awk
checking if awk is broken... no
checking for V8 Javascript Engine... yes, shared
checking for g++... g++
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking how to run the C++ preprocessor... g++ -E
checking for V8 version... NONE
configure: error: could not determine libv8 version

➜  v8  pwd
/Users/fro/build/v8
➜  v8  tree
.
├── include
│   ├── v8-debug.h
│   ├── v8-platform.h
│   ├── v8-profiler.h
│   ├── v8-testing.h
│   ├── v8-util.h
│   ├── v8.h
│   └── v8config.h
└── lib
    ├── libicui18n.dylib
    ├── libicuuc.dylib
    └── libv8.dylib

2 directories, 10 files
```

Why it "could not determine libv8 version"? Thanks in advance!

Same problem here (Yosemite 10.10.2). I have overridden this by adding dirty hack to my config.m4 file (around line 50):
```
if (testfile.is_open()) {
        //testfile << v8::V8::GetVersion();
        testfile << "3.30.33.16";
        testfile << "\n";
        testfile.close();
        return 0;
}
```

then `` ./configure --with-v8js=/tmp/v8-install CXXFLAGS="-D_GLIBCXX_USE_NANOSLEEP"`` ends correctly and creates Makefile.

But when I call 'make' I get couple of theese errors

```
/usr/local/Cellar/php56/5.6.5/include/php/Zend/zend_API.h:112:53: note: expanded from macro 'ZEND_BEGIN_ARG_INFO'
        ZEND_BEGIN_ARG_INFO_EX(name, 0, ZEND_RETURN_VALUE, -1)
                                                           ^~
/usr/local/Cellar/php56/5.6.5/include/php/Zend/zend_API.h:110:20: note: expanded from macro 'ZEND_BEGIN_ARG_INFO_EX'
                { NULL, 0, NULL, required_num_args, 0, return_reference, 0, 0 },
                                 ^
/tmp/v8js/v8js_class.cc:1039:1: error: constant expression evaluates to -1 which cannot be narrowed to type 'zend_uint' (aka 'unsigned int') [-Wc++11-narrowing]
ZEND_BEGIN_ARG_INFO(arginfo_v8js_getextensions, 0)
```

So I guess there is still something wrong...

I'm also on OSX 10.10.2 and can't build. Tried @jaromirnyklicek approach of editing config.m4, but I still receive:

> configure: error: could not determine libv8 version

I've finally fetched a MacOS vagrant image from the internet to give it a try myself; I'm usually on just on GNU/Linux boxes so bear with me...

The rpath stuff doesn't seem to be supported by Darwin, hence I've prepared a patched config.m4 file I'll eventually merge.

I used the Apple LLVM 6.0 compiler (clang); in combination with v8 *3.28.71.19* (latest one from the 3.28 line) *it works* for me.  I also got the error @jaromirnyklicek  mentioned

```
/tmp/v8js/v8js_class.cc:1039:1: error: constant expression evaluates to -1 which cannot be narrowed to type 'zend_uint' (aka 'unsigned int') [-Wc++11-narrowing]
ZEND_BEGIN_ARG_INFO(arginfo_v8js_getextensions, 0)
```
... which is caused by PHP internal macros, which cast -1 to unsigned integer.  gcc accepts that, but clang doesn't.  However you can pass `CXXFLAGS="-Wno-c++11-narrowing"` to configure.

----

v8 versions 3.29.36 or later do **not** currently work.  This is because v8 needs a library named libplatform there, which is available but it uses symbols from v8 library that are not externally visible and hence cause runtime problems.

* v8js.so needs symbol `__ZN2v84base5MutexC1Ev` (via statically linked libplatform); capital U for "undefined"
```
vagrant-osx:v8js vagrant$ nm /usr/local/lib/php/extensions/no-debug-non-zts-20131226/v8js.so | grep __ZN2v84base5MutexC1Ev
                 U __ZN2v84base5MutexC1Ev
```
* libv8.dylib does even define it ("t" for text segment aka code), but it's a lower case letter (which means internal)
```
vagrant-osx:v8js vagrant$ nm /usr/local/lib/libv8.dylib | grep __ZN2v84base5MutexC1Ev
00000000003447de t __ZN2v84base5MutexC1Ev
```

* on a GNU/Linux system this symbol is public (capital T)
```
root@97c2d1abbaf4:/# nm /usr/local/v8/lib/libv8.so  | grep -ie ZN2v84base5MutexC1Ev
0000000000677170 T _ZN2v84base5MutexC1Ev
```
... and hence it's all fine there

----

TL;DR
* v8 versions from the 3.28 line supposed to work fine
* need to dig further regarding how to externalize the above symbol on darwin
* I have a draft README.md file for MacOS around and will merge it with the config.m4 patch soonish

cheers
  ~stesie


by the way, I think you all don't need the `CXXFLAGS="-D_GLIBCXX_USE_NANOSLEEP"` flags, they are just a work around needed for gcc 4.7

--- a/README.Linux.md
+++ b/README.Linux.md
@@ -1,0 +1,63 @@
+V8Js on GNU/Linux
+=================
+
+Installation of V8Js on GNU/Linux is pretty much straight forward.
+
+The biggest hurdle actually is that you need a rather new v8 library.
+However most distributions still ship the rusty version 3.14, publish
+years ago, since Node.js requires such an old version.
+
+This means that you usually need to compile v8 on your own before
+you can start to compile & install v8js itself.
+
+Compile latest v8
+-----------------
+
+```
+cd /tmp
+
+# Install depot_tools first (needed for source checkout)
+git clone https://chromium.googlesource.com/chromium/tools/depot_tools.git
+export PATH=`pwd`/depot_tools:"$PATH"
+
+# Download v8
+fetch v8
+cd v8
+
+# (optional) If you'd like to build a certain version:
+git checkout 3.32.6
+gclient sync
+
+# Build
+make native library=shared -j8
+
+# Install to /usr
+sudo mkdir -p /usr/lib /usr/include
+sudo cp out/native/lib.target/lib*.so /usr/lib/
+sudo cp -R include/* /usr/include
+echo -e "create /usr/lib/libv8_libplatform.a\naddlib out/native/obj.target/tools/gyp/libv8_libplatform.a\nsave\nend" | sudo ar -M
+```
+
+If you don't want to overwrite the system copy of v8, replace `/usr` in
+the above commands with `/tmp/v8-install` and then add
+`--with-v8js=/tmp/v8-install` to the php-v8js `./configure` command below.
+
+`libv8_libplatform.a` should not be copied directly since it's a thin
+archive, i.e. it contains only pointers to the build objects, which
+otherwise must not be deleted.  The simple mri-script converts the
+thin archive to a normal archive.
+
+
+Compile php-v8js itself
+-----------------------
+
+```
+cd /tmp
+git clone https://github.com/preillyme/v8js.git
+cd v8js
+phpize
+./configure
+make
+make test
+sudo make install
+```

--- a/README.MacOS.md
+++ b/README.MacOS.md
@@ -1,0 +1,79 @@
+V8Js on MacOS
+=============
+
+Installation of V8Js on MacOS is pretty much straight forward.
+
+First of all you need a pretty fresh installation of v8 library.
+If you have brew around, just `brew install v8` and you should be done.
+Otherwise you need to compile latest v8 manually.
+
+Compile latest v8
+-----------------
+
+```
+cd /tmp
+
+# Install depot_tools first (needed for source checkout)
+git clone https://chromium.googlesource.com/chromium/tools/depot_tools.git
+export PATH=`pwd`/depot_tools:"$PATH"
+
+# Download v8
+fetch v8
+cd v8
+
+# (optional) If you'd like to build a certain version:
+git checkout 3.32.6
+gclient sync
+
+# Build
+make native library=shared -j8
+
+# Install to /usr
+mkdir -p /usr/local/lib /usr/local/include
+cp out/native/lib*.dylib /usr/local/lib/
+cp out/native/libv8_*.a  /usr/local/lib/
+cp -R include/* /usr/local/include
+```
+
+You cannot install the libraries to any location you want since they
+have a install name that is baked into the library.  You can check
+the install name with `otool -D out/native/libv8.dylib`.
+
+During the build snapshot generation may fail like so:
+
+```
+ACTION tools_gyp_v8_gyp_v8_snapshot_target_run_mksnapshot /Users/vagrant/v8/out/native/obj.target/v8_snapshot/geni/snapshot.cc
+dyld: Library not loaded: /usr/local/lib/libicui18n.dylib
+  Referenced from: /Users/vagrant/v8/out/native/mksnapshot
+  Reason: image not found
+/bin/sh: line 1: 18964 Trace/BPT trap: 5       "/Users/vagrant/v8/out/native/mksnapshot" --log-snapshot-positions --logfile "/Users/vagrant/v8/out/native/obj.target/v8_snapshot/geni/snapshot.log" --random-seed 314159265 "/Users/vagrant/v8/out/native/obj.target/v8_snapshot/geni/snapshot.cc"
+make[1]: *** [/Users/vagrant/v8/out/native/obj.target/v8_snapshot/geni/snapshot.cc] Error 133
+make: *** [native] Error 2
+```
+
+... if that happens, just copy libicu*.dylib to /usr/local/lib already
+and start make again (then simply continue with installation):
+
+```
+cp out/native/libicu*.dylib /usr/local/lib/
+make native library=shared -j8
+```
+
+
+Compile php-v8js itself
+-----------------------
+
+If you're using Apple LLVM compiler (instead of gcc) you need to pass the `-Wno-c++11-narrowing`
+flag.  Otherwise compilation fails due to narrowing errors in PHP itself, which LLVM is much pickier
+with (compared to gcc).
+
+```
+cd /tmp
+git clone https://github.com/preillyme/v8js.git
+cd v8js
+phpize
+./configure CXXFLAGS="-Wno-c++11-narrowing"
+make
+make test
+make install
+```

--- a/README.md
+++ b/README.md
@@ -23,58 +23,22 @@ Minimum requirements
 
   This embedded implementation of the V8 engine uses thread locking so it should work with ZTS enabled.
   However, this has not been tested yet.
-  
-COMPILING LATEST VERSION
-========================
 
-Instead of compiling manually you might want to pull from the [V8Js docker
-repository](https://registry.hub.docker.com/u/stesie/v8js/).
+
+Compiling latest version
+------------------------
+
+For some very first steps, instead of compiling manually you might want to try out the [V8Js docker
+image](https://registry.hub.docker.com/u/stesie/v8js/).  It has v8, v8js and php-cli pre-installed
+so you can give it a try with PHP in "interactive mode".  There is no Apache, etc. running however.
 
 You also might want to try the Debian & Ubuntu packages available from
 the Jenkins site at https://jenkins.brokenpipe.de/
 
-Building on Microsoft Windows is more complicated, see README.Win32.md file
-for a quick run through.
-
-Compile latest v8
------------------
-
-```
-cd /tmp
-git clone https://chromium.googlesource.com/chromium/tools/depot_tools.git
-export PATH=`pwd`/depot_tools:"$PATH"
-git clone https://github.com/v8/v8.git
-cd v8
-make dependencies
-make native library=shared -j8
-sudo mkdir -p /usr/lib /usr/include
-sudo cp out/native/lib.target/lib*.so /usr/lib/
-echo -e "create /usr/lib/libv8_libplatform.a\naddlib out/native/obj.target/tools/gyp/libv8_libplatform.a\nsave\nend" | sudo ar -M
-sudo cp -R include/* /usr/include
-
-```
-
-If you don't want to overwrite the system copy of v8, replace `/usr` in
-the above commands with `/tmp/v8-install` and then add
-`--with-v8js=/tmp/v8-install` to the php-v8js `./configure` command below.
-
-`libv8_libplatform.a` should not be copied directly since it's a thin
-archive, i.e. it contains only pointers to  the build objects, which
-otherwise must not be deleted.  The simple mri-script converts the
-thin archive to a normal archive.
-
-Compile php-v8js itself
------------------------
-
-```
-cd /tmp
-git clone https://github.com/preillyme/v8js.git
-cd v8js
-phpize
-./configure
-make
-sudo make install
-```
+Building on Microsoft Windows is a bit more involved, see README.Win32.md file
+for a quick run through.  Building on GNU/Linux and MacOS X is straight forward,
+see README.Linux.md and README.MacOS.md files for a walk through with platform
+specific notes.
 
 
 PHP API

--- a/config.m4
+++ b/config.m4
@@ -110,22 +110,32 @@ AC_LINK_IFELSE([AC_LANG_PROGRAM([[#include <v8-debug.h>]],
     dnl link in libplatform to make our life easier.
     PHP_ADD_INCLUDE($V8_DIR)
 
-    SEARCH_FOR="lib/libv8_libplatform.a"
-    AC_MSG_CHECKING([for libv8_libplatform.a])
+    case $host_os in
+      darwin* )
+        static_link_extra="libv8_libplatform.a libv8_libbase.a"
+        ;;
+      * )
+        static_link_extra="libv8_libplatform.a"
+        ;;
+    esac
 
-    for i in $PHP_V8JS $SEARCH_PATH ; do
-      if test -r $i/$SEARCH_FOR; then
-        LIBPLATFORM_DIR=$i
-        AC_MSG_RESULT(found in $i)
+    for static_link_extra_file in $static_link_extra; do
+      AC_MSG_CHECKING([for $static_link_extra_file])
+
+      for i in $PHP_V8JS $SEARCH_PATH ; do
+        if test -r $i/lib/$static_link_extra_file; then
+          static_link_dir=$i
+          AC_MSG_RESULT(found in $i)
+        fi
+      done
+
+      if test -z "$static_link_dir"; then
+        AC_MSG_RESULT([not found])
+        AC_MSG_ERROR([Please provide $static_link_extra_file next to the libv8.so, see README.md for details])
       fi
+
+      LDFLAGS="$LDFLAGS $static_link_dir/lib/$static_link_extra_file"
     done
-
-    if test -z "$LIBPLATFORM_DIR"; then
-      AC_MSG_RESULT([not found])
-      AC_MSG_ERROR([Please provide libv8_libplatform.a next to the libv8.so, see README.md for details])
-    fi
-
-    LDFLAGS="$LDFLAGS $LIBPLATFORM_DIR/$SEARCH_FOR"
   fi
 
   AC_CACHE_CHECK(for C standard version, ac_cv_v8_cstd, [

--- a/config.m4
+++ b/config.m4
@@ -6,7 +6,14 @@ if test "$PHP_V8JS" != "no"; then
   SEARCH_FOR="include/v8.h"
   
   if test -r $PHP_V8JS/$SEARCH_FOR; then
-    LDFLAGS="$LDFLAGS -Wl,--rpath=$PHP_V8JS/$PHP_LIBDIR"
+    case $host_os in
+      darwin* )
+        # MacOS does not support --rpath
+        ;;
+      * )
+        LDFLAGS="$LDFLAGS -Wl,--rpath=$PHP_V8JS/$PHP_LIBDIR"
+        ;;
+    esac
     V8_DIR=$PHP_V8JS
   else
     AC_MSG_CHECKING([for V8 files in default path])
@@ -31,7 +38,17 @@ if test "$PHP_V8JS" != "no"; then
   old_LIBS=$LIBS
   old_LDFLAGS=$LDFLAGS
   old_CPPFLAGS=$CPPFLAGS
-  LDFLAGS="-Wl,--rpath=$V8_DIR/$PHP_LIBDIR -L$V8_DIR/$PHP_LIBDIR"
+
+  case $host_os in
+    darwin* )
+      # MacOS does not support --rpath
+      LDFLAGS="-L$V8_DIR/$PHP_LIBDIR"
+      ;;
+    * )
+      LDFLAGS="-Wl,--rpath=$V8_DIR/$PHP_LIBDIR -L$V8_DIR/$PHP_LIBDIR"
+      ;;
+  esac
+
   LIBS=-lv8
   CPPFLAGS=-I$V8_DIR/include
   AC_LANG_SAVE

